sql/src/main/java/io/crate/executor/transport/distributed/BroadcastDistributingDownstream.java

```diff
@@ -0,0 +1,88 @@
+/*
+ * Licensed to CRATE.IO GmbH ("Crate") under one or more contributor
+ * license agreements.  See the NOTICE file distributed with this work for
+ * additional information regarding copyright ownership.  Crate licenses
+ * this file to you under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.  You may
+ * obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.  See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ *
+ * However, if you have executed another commercial license agreement
+ * with Crate these terms will supersede the license and you may use the
+ * software solely pursuant to the terms of the relevant commercial agreement.
+ */
+
+package io.crate.executor.transport.distributed;
+
+import io.crate.Constants;
+import io.crate.Streamer;
+import io.crate.core.collections.Bucket;
+import io.crate.core.collections.Row;
+import org.elasticsearch.common.logging.ESLogger;
+import org.elasticsearch.common.logging.Loggers;
+
+import java.util.Collection;
+import java.util.UUID;
+
+public class BroadcastDistributingDownstream extends DistributingDownstream {
+
+    private static final ESLogger LOGGER = Loggers.getLogger(BroadcastDistributingDownstream.class);
+
+    private final SingleBucketBuilder bucketBuilder;
+
+    public BroadcastDistributingDownstream(UUID jobId,
+                                           int targetExecutionNodeId,
+                                           int bucketIdx,
+                                           Collection<String> downstreamNodeIds,
+                                           TransportDistributedResultAction transportDistributedResultAction,
+                                           Streamer<?>[] streamers) {
+        super(jobId, targetExecutionNodeId, bucketIdx, downstreamNodeIds, transportDistributedResultAction, streamers);
+
+        bucketBuilder = new SingleBucketBuilder(streamers);
+    }
+
+    @Override
+    public boolean setNextRow(Row row) {
+        if (allDownstreamsFinished()) {
+            return false;
+        }
+        try {
+            bucketBuilder.setNextRow(row);
+            sendRequestsIfNeeded();
+        } catch (Exception e) {
+            fail(e);
+            return false;
+        }
+        return true;
+    }
+
+    private void sendRequestsIfNeeded() {
```

Comment:
could lead to a race condition, since setNextRow can be called from multiple threads